diff --git a/kn5toac.cpp b/kn5toac.cpp
index d0d1ba4..019214e 100644
--- a/kn5toac.cpp
+++ b/kn5toac.cpp
@@ -55,6 +55,9 @@ static void writeConfig(const std::filesystem::path& inputPath, const std::strin
     ini suspensions(inputPathString + "data/suspensions.ini");
     ini tires(inputPathString + "data/tyres.ini");
 
+    std::array<float,3>   graphicsOffset = car.getFloatArray3Value("BASIC", "GRAPHICS_OFFSET");
+    float pitchDegrees = car.getFloatValue("BASIC", "GRAPHICS_PITCH_ROTATION");
+
     std::string modelFileName = inputPath.filename().string() + ".ac";
 
     std::ofstream   fout(filename);
@@ -107,12 +110,13 @@ static void writeConfig(const std::filesystem::path& inputPath, const std::strin
     fout << "\t\t</section>" << std::endl;
     fout << "\t\t<section name=\"Steer Wheel\">" << std::endl;
     fout << "\t\t\t<attstr name=\"model\" val=\"steer.acc\"/>" << std::endl;
-    fout << "\t\t\t<attstr name=\"hi res model\" val=\"histeer.acc\"/>" << std::endl;
-    kn5::Vec3   steer = { 0, 0, 0 };
+    kn5::Node* histeerNode = model.findNode(kn5::Node::Transform, "STEER_HR");
+    if (histeerNode)
+        fout << "\t\t\t<attstr name=\"hi res model\" val=\"histeer.acc\"/>" << std::endl;
     kn5::Node* steerNode = model.findNode(kn5::Node::Transform, "STEER_LR");
-
     if (steerNode)
     {
+        kn5::Vec3   steer = { 0, 0, 0 };
         steer[0] = steerNode->m_matrix.m_data[3][2];
         steer[1] = steerNode->m_matrix.m_data[3][0];
         steer[2] = steerNode->m_matrix.m_data[3][1];
@@ -120,8 +124,8 @@ static void writeConfig(const std::filesystem::path& inputPath, const std::strin
         fout << "\t\t\t<attnum name=\"xpos\" val=\"" << steer[0] << "\"/>" << std::endl;
         fout << "\t\t\t<attnum name=\"ypos\" val=\"" << steer[1] << "\"/>" << std::endl;
         fout << "\t\t\t<attnum name=\"zpos\" val=\"" << steer[2] << "\"/>" << std::endl;
+//	    fout << "\t\t\t<attnum name=\"angle\" val=\"" << 0 << "\"/>" << std::endl;
     }
-//	<attnum name="angle" val="0"/>
     fout << "\t\t</section>" << std::endl;
     fout << "\t\t<section name=\"Driver\">" << std::endl;
     fout << "\t\t</section>" << std::endl;
@@ -684,6 +688,14 @@ int main(int argc, char* argv[])
         xform.m_data[3][2] = 0;
         xform.m_data[3][3] = 1;
 
+        std::filesystem::path   carPath = inputPath;
+
+        carPath.append("data/car.ini");
+
+        ini car(carPath.string());
+
+        std::array<float,3>   graphicsOffset = car.getFloatArray3Value("BASIC", "GRAPHICS_OFFSET");
+
         if (extractCarParts)
         {
             std::filesystem::path extractFilePath = outputPath;
