diff --git a/CMakeLists.txt b/CMakeLists.txt
index 73e4248..5ed323a 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -4,7 +4,7 @@ project (kn5toac VERSION 0.1 LANGUAGES CXX)
 
 configure_file(config.h.in config.h)
 
-add_executable(kn5toac kn5toac.cpp kn5.h kn5.cpp ini.h ini.cpp lut.h lut.cpp)
+add_executable(kn5toac kn5toac.cpp kn5.h kn5.cpp ini.h ini.cpp lut.h lut.cpp acd.h acd.cpp)
 
 target_compile_features(kn5toac PUBLIC cxx_std_17)
 target_include_directories(kn5toac PUBLIC "${PROJECT_BINARY_DIR}")
diff --git a/kn5.h b/kn5.h
index 3c73698..cf0bd96 100644
--- a/kn5.h
+++ b/kn5.h
@@ -225,7 +225,7 @@ public:
     void getUsedMaterials(const kn5::Node& node, std::set<int>& usedMaterialIDs) const;
     int getNewMaterialID(int materialID, const std::set<int>& usedMaterialIDs) const;
 
-    Node * kn5::findNode(Node& node, Node::NodeType type, const std::string& name);
+    Node * findNode(Node& node, Node::NodeType type, const std::string& name);
 
     int32_t                 m_version = 0;
     int32_t                 m_unknown = 0;
diff --git a/kn5toac.cpp b/kn5toac.cpp
index 7dd5cda..2e5c2c9 100644
--- a/kn5toac.cpp
+++ b/kn5toac.cpp
@@ -246,7 +246,7 @@ static void writeConfig(const std::filesystem::path& inputPath, const std::strin
     std::string traction = drivetrain.getValue("TRACTION", "TYPE");
     if (traction == "AWD")
         traction = "4WD";
-       fout << "\t\t<attstr name=\"type\" val=\"" << traction << "\"/>" << std::endl;
+    fout << "\t\t<attstr name=\"type\" val=\"" << traction << "\"/>" << std::endl;
 //     <attnum name="inertia" unit="kg.m2" val="0.0091"/>
     fout << "\t</section>" << std::endl;
 
@@ -771,7 +771,7 @@ int main(int argc, char* argv[])
         }
     }
 
-    if (writeConfig)
+    if (writeCarConfig)
     {
         std::filesystem::path dataDirectory = outputPath;
 

